src/SprykerSdk/AsyncApi/AsyncApi/Loader/AsyncApiLoader.php

```diff
@@ -166,16 +166,94 @@ protected function formatMessageFromOneOf(array $messages, array $asyncApi): arr
         foreach ($messages as $message) {
             foreach ($message as $reference) {
                 $messageName = $this->resolveReferenceKey($reference);
-                $resolvedMessage = $this->resolveReference($asyncApi, $reference);
 
-                $messageWithResolvedReferences = $this->resolveReferences($resolvedMessage, $asyncApi);
-                $formattedMessages[$messageName] = $messageWithResolvedReferences;
+                $formattedMessages[$messageName] = $this->resolveWithResolvedReferences($asyncApi, $reference);
             }
         }
 
         return $formattedMessages;
     }
 
+    /**
+     * @param array $asyncApi
+     * @param string $reference
+     *
+     * @return array<mixed>
+     */
+    protected function resolveWithResolvedReferences(array $asyncApi, string $reference): array
+    {
+        $resolved = $this->resolveReference($asyncApi, $reference);
+
+        if (!$resolved) {
+            return $this->resolveWithResolvedReferencesFromRemoteApi($asyncApi, $reference);
+        }
+
+        return $this->resolveReferences($resolved, $asyncApi);
+    }
+
+    /**
+     * @param array $asyncApi
+     * @param string $reference
+     *
+     * @return array<mixed>
+     */
+    protected function resolveWithResolvedReferencesFromRemoteApi(array $asyncApi, string $reference): array
+    {
+        [$value, $remoteReference] = $this->getReferenceInRootAsyncApi($asyncApi, $reference);
+
+        $key = array_key_first($value);
+        $value = $value[$key];
+
+        $valueFragments = explode('#', $value);
+        $value = $valueFragments[0];
+
+        if (!filter_var($value, FILTER_VALIDATE_URL)) {
+            return [];
+        }
+
+        if (isset($valueFragments[1])) {
+            // Ensure that if the prefix has a trailing slash we only have one at the end.
+            $remoteReferencePrefix = rtrim($valueFragments[1], '/') . '/';
+            $remoteReference = $remoteReferencePrefix . $remoteReference;
+        }
+
+        $remoteReference = sprintf('#/%s', ltrim($remoteReference, '#/'));
+
+        $remoteAsyncApi = $this->getYmlArray($value);
```

Comment:
it would be good to log a warning when the remote URL responds with `4xx`, it can help us and other developers to find wrongly provided URLs.

```php
if (($data = @file_get_contents($remoteUrl)) === false) {
      $error = error_get_last();
      echo "HTTP request failed. Error was: " . $error['message'];
} else {
      echo "Everything went better than expected";
}
```

Reply:
Good one. I added a test case for this an the needed code.